### PR TITLE
Wrap Commit result and fetch from the command line

### PR DESCRIPTION
**Pull request by DefaultRyan**
## Summary of the pull request
Use the git command line to fetch the latest commit for a file.

## Detailed description of the pull request / Additional comments
In GVFS-enabled repos, while we can walk the entire commit log, parts of it are virualized out and Libgit2 doesn't cause them to appear.
## PR checklist
- [ ] Closes #xxx
- [ ] Tests added/passed
- [ ] Documentation updated

### CODE REVIEW

File: extensions/GitExtension/FileExplorerGitIntegration/Models/CommitLogCache.cs [hunk 1]
```diff
@@ -17,32 +17,145 @@ namespace FileExplorerGitIntegration.Models;
...
+        var parts = result.Output.Split('\n');
+        string message = parts[0];
+        string authorName = parts[1];
+        string authorEmail = parts[2];
+        DateTimeOffset authorWhen = DateTimeOffset.Parse(parts[3], null, System.Globalization.DateTimeStyles.RoundtripKind);
+        string sha = parts[4];
+        return new CommitWrapper(message, authorName, authorEmail, authorWhen, sha);
+    }
+
+    private CommitWrapper? FindLastCommitUsingLibGit2Sharp(string relativePath)
```
Comment (DefaultRyan) [inline extensions/GitExtension/FileExplorerGitIntegration/Models/CommitLogCache.cs:112]: This was moved verbatim from RepositoryWrapper

File: extensions/GitExtension/FileExplorerGitIntegration/Models/GitLocalRepository.cs [hunk 2]
```diff
@@ -72,89 +72,54 @@ IPropertySet ILocalRepository.GetProperties(string[] properties, string relative
             switch (propName)
             {
                 case "System.VersionControl.LastChangeMessage":
-                    if (latestCommit == null)
+                    latestCommit ??= FindLatestCommit(relativePath, repository);
+                    if (latestCommit != null)
```
Comment (dkbennett) [inline extensions/GitExtension/FileExplorerGitIntegration/Models/GitLocalRepository.cs:76]: Nit: use "is not null" which shortcuts overloaded operators for the null case. Same comment throughout for other null comparisons.

File: extensions/GitExtension/FileExplorerGitIntegration/Models/CommitLogCache.cs [hunk 3]
```diff
@@ -17,32 +17,145 @@ namespace FileExplorerGitIntegration.Models;
...
+
+        return result;
     }
 
-    IEnumerator IEnumerable.GetEnumerator()
+    private CommitWrapper? FindLastCommitUsingCommandLine(string relativePath)
     {
-        return GetEnumerator();
+        var result = GitExecute.ExecuteGitCommand(_gitDetect.GitConfiguration.ReadInstallPath(), _workingDirectory, $"log -n 1 --pretty=format:%s%n%an%n%ae%n%aI%n%H -- {relativePath}");
+        if (result.Status != Microsoft.Windows.DevHome.SDK.ProviderOperationStatus.Success || result.Output == null)
```
Comment (dkbennett) [inline extensions/GitExtension/FileExplorerGitIntegration/Models/CommitLogCache.cs:98]: Nit: parentheses around each condition

**Review by dkbennett — APPROVED** [review]

**Review by ssparach — APPROVED** [review]

File: extensions/GitExtension/FileExplorerGitIntegration/Models/GitLocalRepository.cs (hunk 2 above)
Comment (DefaultRyan) [reply extensions/GitExtension/FileExplorerGitIntegration/Models/GitLocalRepository.cs:76]: > Nit: use "is not null" which shortcuts overloaded operators for the null case. Same comment throughout for other null comparisons.

Wow! Today I Learned...

**Pull request merged**